CHANGELOG.md

```diff
@@ -33,6 +34,19 @@
 - [v0.2.0](#v020)
 - [v0.1.0](#v010)
 
+## [v1.8.3]
+
+> Release date: 2021/12/02
+
+### Fixed
+
+- `--select-tag` applies its tags to newly-created entities whether or not the
+  tag is also present in the state file metadata.
+  [#517](https://github.com/Kong/deck/pull/517)
+- Timeouts in `Syncer.Run()` now return an error. Downstream clients using decK
+  as a library can determine when their sync attempt failed due to a timeout.
+  [#529](https://github.com/Kong/deck/pull/529)
```

Comment:
https://github.com/Kong/deck/commit/cbd958a5cc3fb3f68c354de493a28487de348a3f to further explain what "failing silently" meant.

This is a bit murkier because there's an error where there wasn't before, but the successful return was a lie and indistinguishable from actual success based on `Syncher.Run()` alone (it returns nothing other than errors). You could possibly have checked your context to see if it had expired (still works, just no longer necessary) or independently verified the updated objects (extremely painful), both of which are a bit :thinking: if someone put them in place and didn't report an issue.

Actually relying on partial syncs seems unrealistic, since there's no way to control when in the sync process the timeout actually occurs: the subset of changes applied is essentially random, though it is ordered (deletes always come after updates and so on).

In practice, error handling should be one of:
- You ignore errors form `Syncher.Run()` entirely, and this change has no impact on you.
- You handle errors generically. There's an additional condition where you'll see an error, but arguably one where you do indeed want to see the error, because the requested action didn't complete successfully.
- You handle specific errors differently. There's now a new type of error you need to handle if you want to take some specific action based on it, but you were either ignoring unknown errors or had a catch-all case at the end.

This should furthermore affect library uses of deck only: the CLI can only pass in a command context, and that only exits when you quit. There is a cosmetic change since we [print the previously silenced error](https://github.com/Kong/deck/pull/529#issue-1061656372), but any screen-scraping error handling should still fall into one of the above categories. Library usage is currently [a bunch of KIC forks](https://github.com/Kong/deck/network/dependents?dependent_type=REPOSITORY), so there's not really anything for us to worry about there.

IMO based on that it falls on the non-breaking fix side of things. In the world of infinite possibilities we're potentially wrong, but based on what we know, those are very edgy edge cases if they exist, to the point that we'd need to discuss those problems with those individual users rather than being able to make a blanket "this will break your stuff, you must make changes" statement.

Reply:
> In the world of infinite possibilities we're potentially wrong, but based on what we know, those are very edgy edge cases if they exist, to the point that we'd need to discuss those problems with those individual users rather than being able to make a blanket "this will break your stuff, you must make changes" statement.

I get your point but I disagree with this sentiment. The central question to me is: "are we making an interface change here?" and the answer seems to be "yes". Whether this interface change is benign (or even desirable) is another question, but (to me) a strictly less important one than whether the interface change is there in the first place.

Your impact analysis (_This should furthermore affect library uses of deck only_) indicates that the blast radius of this decision is minimal, so (while I generally expect that dependable parts of APIs remain fully backwards-compatible between patch versions), I don't think further investment into this specific versioning decision is worthwhile. That said, I do think that (going forward) API versioning deserves a stricter approach to semantic differences between versions.